src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -132,23 +156,73 @@ private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax
         {
             // symbol and parameters for Deconstruct
             ImmutableArray<BoundDeconstructValuePlaceholder> outPlaceholders;
-            var deconstructInvocation = MakeDeconstructInvocationExpression(variables.Length, targetPlaceholder, syntax, diagnostics, out outPlaceholders);
-
+            var deconstructInvocation = MakeDeconstructInvocationExpression(variables.Count, targetPlaceholder, syntax, diagnostics, out outPlaceholders);
             if (deconstructInvocation.HasAnyErrors)
             {
                 return null;
             }
-            else
+
+            SetInferredTypes(variables, outPlaceholders.SelectAsArray(p => p.Type));
+
+            return new BoundDeconstructionDeconstructStep(syntax, deconstructInvocation, targetPlaceholder, outPlaceholders);
+        }
+
+        /// <summary>
+        /// Inform the variables about found types (whether one was found or not).
+        /// </summary>
+        static private void SetInferredTypes(ArrayBuilder<DeconstructionVariable> variables, ImmutableArray<TypeSymbol> foundTypes)
+        {
+            if (!variables.Any(v => IsDeconstructionLocalPendingInference(v)))
+            {
+                return;
+            }
+
+            var matchCount = Math.Min(variables.Count, foundTypes.Length);
+            for (int i = 0; i < matchCount; i++)
             {
-                return new BoundDeconstructionDeconstructStep(syntax, deconstructInvocation, targetPlaceholder, outPlaceholders);
+                var variable = variables[i];
+                if (IsDeconstructionLocalPendingInference(variable))
+                {
+                    var local = ((DeconstructionLocalPendingInference)variable.Single).SetInferredType(foundTypes[i], true);
+                    variables[i] = new DeconstructionVariable(local);
+                }
+            }
+        }
+
+        private static bool IsDeconstructionLocalPendingInference(DeconstructionVariable variable)
+        {
+            return !variable.IsNested && variable.Single.Kind == BoundKind.DeconstructionLocalPendingInference;
+        }
+
+        /// <summary>
+        /// Find any deconstruction locals that are still pending inference and fail their inference.
+        /// </summary>
+        private void FailRemainingInferences(ArrayBuilder<DeconstructionVariable> variables, DiagnosticBag diagnostics)
+        {
+            var count = variables.Count;
+            for (int i = 0; i < count; i++)
+            {
+                var variable = variables[i];
+                if (variable.IsNested)
+                {
+                    FailRemainingInferences(variable.Nested, diagnostics);
+                }
+                else
+                {
+                    if (variable.Single.Kind == BoundKind.DeconstructionLocalPendingInference)
+                    {
+                        var local = ((DeconstructionLocalPendingInference)variable.Single).FailInference(this, diagnostics);
```

Comment:
The primary purpose for failing remaining inferences is to eliminate any remaining instances of `DeconstructionLocalPendingInference`. You're right I don't care as much about reporting additional errors.
I'll remove the error reporting inside `FailInference` to avoid duplicates (as seen in the last test, called `Mismatch`).